### PR TITLE
add sortMaps:bool option

### PR DESCRIPTION
**Pull request by cryptix**
This starts the work on adding `sortMap:bool` to make it easy to create canonical map representations.

I had trouble resolving these two comments:
* https://github.com/kriszyp/cbor-x/issues/4#issuecomment-1591358930
* https://github.com/kriszyp/cbor-x/issues/61#issuecomment-1336570860

I don't really see the point in moving this to a different library. Since the comment on 4 was newer I just went ahead and did it to see how hard it would be.

My rationale for doing it this way would be:

* re-creating a map/object with sorted keys is not very memory efficient either
* it bubbles up the problem to parts of the stack that might not (want to) care about it
* this code already iterates over keys list anyway, so we just need to sort those

This surely needs more tests before it can be merged/released but I wanted to get a temp check before I spent too much time on this.

### CODE REVIEW

File: encode.js [hunk 1]
```diff
@@ -1227,5 +1234,8 @@ export const { NEVER, ALWAYS, DECIMAL_ROUND, DECIMAL_FIT } = FLOAT32_OPTIONS
 export const REUSE_BUFFER_MODE = 512
 export const RESET_BUFFER_MODE = 1024
 export const THROW_ON_ITERABLE = 2048
-
-
+const lexSortFn = (a, b) => {
+	if (a.length < b.length) return -1
```
Comment (kriszyp) [inline encode.js:1238]: Does the CBOR spec say that length takes precedence over bytewise/lexigraphic order? I don't think it does, does it?

File: encode.js [hunk 2]
```diff
@@ -507,11 +508,12 @@ export class Encoder extends Decoder {
 								encode(encoder.encodeKey(key))
 								encode(entryValue)
 							} 
-						} else { 
-							for (let [ key, entryValue ] of value) {
+						} else {
+							const keys = Array.from(value.keys()).sort(lexSortFn)
```
Comment (kriszyp) [inline encode.js:512]: This would still need to the conditional for the flag, right?

File: encode.js [hunk 3]
```diff
@@ -627,16 +629,15 @@ export class Encoder extends Decoder {
 				targetView.setUint32(position, length)
 				position += 4
 			}
-			let key
 			if (encoder.keyMap) { 
 				for (let i = 0; i < length; i++) {
 					encode(encoder.encodeKey(keys[i]))
-					encode(vals[i])
```
Comment (kriszyp) [inline encode.js:634]: I wonder if we could still preserve the values when sortMaps is not enabled, for performance.

**Review by kriszyp — COMMENTED** [review]
Sorry for the slow response. This looks like a very reasonable addition.

File: encode.js (hunk 1 above)
Comment (cryptix) [reply encode.js:1238]: It actually does: https://www.rfc-editor.org/rfc/rfc8949.html#name-length-first-map-key-orderi

Comment (kriszyp) [reply encode.js:1238]: Ah, ok, I was looking at the previous RFC I guess. This makes sense, and is probably faster.

**Comment by odama626** [comment]
I just came across this, and I think this is exactly what I need for using cbor with signatures. is there anyway I can help get this merged?

**Comment by cryptix** [comment]
> is there anyway I can help get this merged?

Tests would surely be nice 😁


We used this patch a bunch to compare hashes with a go implementation. It should roughly be right but there might be edge cases.

File: encode.js (hunk 3 above)
Comment (odama626) [reply encode.js:634]: I'm guessing you're trying to avoid lookups for each key for performance?

could something like this suffice?

```js
const entries = Object.entries(object);

if (sortMap) {
  entries.sort(lexSort)
}

```

File: encode.js (hunk 2 above)
Comment (odama626) [reply encode.js:512]: Yeah, I think you're right here. this always sorts the keys if there isn't a keymap
